### PR TITLE
reporter: Implement custom how-to-fix texts for `OrtIssue`s

### PR DESCRIPTION
**Pull request by fviernau**
Add a (reporter) command line option to inject custom how-to-fix Markdown texts for OrtIssues from within a Kotlin Script.
Display these injected how-to-fix texts in the `static HTML report` to show that it works.

The integration into the `WebApp` is going to be done in a following PR.

See also: #2347.

### CODE REVIEW

File: examples/how-to-fix-text-provider.kts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+object : HowToFixTextProvider {
+    fun OrtIssue.matchesMessage(pattern: String): Boolean = pattern.toRegex().matches(message)
+
+    override fun getHowToFixText(issue: OrtIssue): String? {
+        // How-to-fix instruction Markdown for scan timeout errors.
+        if (issue.matchesMessage("ERROR: Timeout after .* seconds while scanning file.*")) {
+            return """
+                | To fix this issue please proceede with the following steps
```
Comment (mnonnenmacher) [inline examples/how-to-fix-text-provider.kts:8]: "proceed~~e~~", colon after "steps"-

File: cli/src/funTest/kotlin/ExamplesTest.kt [hunk 2]
```diff
@@ -117,6 +118,14 @@ class ExamplesTest : StringSpec() {
             }
         }
 
+        "how-to-fix-text-provider.kts can be run" {
+            val script = takeExampleFile("how-to-fix-text-provider.kts").readText()
+
+            HowToFixTextProvider.fromKotlinScript(script)
+
+            // TODO: It should also be verified that the script works as expected.
```
Comment (mnonnenmacher) [inline cli/src/funTest/kotlin/ExamplesTest.kt:126]: Yes, please do, it's easy to do, isn't it?

File: examples/how-to-fix-text-provider.kts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+object : HowToFixTextProvider {
+    fun OrtIssue.matchesMessage(pattern: String): Boolean = pattern.toRegex().matches(message)
```
Comment (mnonnenmacher) [inline examples/how-to-fix-text-provider.kts:2]: Wouldn't it be useful to have this function as part of the interface? I assume most how-to-fix texts will be triggered by matching a message.

File: reporter/src/main/kotlin/HowToFixTextProvider.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.ossreviewtoolkit.model.OrtIssue
+
+/**
+ * Resolves how-to-fix texts in Markdown format for any given [OrtIssue].
+ */
+interface HowToFixTextProvider {
+    /**
+     * Return a Markdown text describing how to fix the given [issue]. Non-null return values override the default
+     * how-to-fix texts, while a null value keeps the default.
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/HowToFixTextProvider.kt:44]: What default? Do you plan to add default how to fix messages?

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: reporter/src/main/kotlin/HowToFixTextProvider.kt (hunk 4 above)
Comment (fviernau) [reply reporter/src/main/kotlin/HowToFixTextProvider.kt:44]: They are empty by default currently, but that could be changed.
When it comes to defining the actual how-to-fix texts, I believe it was you who proposed to add as much as possible to ORT. If that was done the default would no more be empty.

File: examples/how-to-fix-text-provider.kts (hunk 3 above)
Comment (fviernau) [reply examples/how-to-fix-text-provider.kts:2]: I intentionally did not add it here yet. My plan was to implement a few custom how-to-fix messages first and then, when these helper functions matured add them to ORT.

File: reporter/src/main/kotlin/HowToFixTextProvider.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.reporter
+
+import org.ossreviewtoolkit.model.OrtIssue
+
+/**
+ * Resolves how-to-fix texts in Markdown format for any given [OrtIssue].
```
[the remark this answers is not in the document]
Comment (fviernau) [reply reporter/src/main/kotlin/HowToFixTextProvider.kt:26]: Done.

File: reporter/src/main/kotlin/HowToFixTextProvider.kt [hunk 6]
```diff
@@ -25,6 +25,15 @@ import org.ossreviewtoolkit.model.OrtIssue
  * Resolves how-to-fix texts in Markdown format for any given [OrtIssue].
  */
 interface HowToFixTextProvider {
+    companion object {
```
[the remark this answers is not in the document]
Comment (fviernau) [reply reporter/src/main/kotlin/HowToFixTextProvider.kt:29]: Done.

File: cli/src/main/kotlin/commands/ReporterCommand.kt [hunk 7]
```diff
@@ -50,6 +50,7 @@ import org.ossreviewtoolkit.model.licenses.orEmpty
 import org.ossreviewtoolkit.model.readValue
 import org.ossreviewtoolkit.model.utils.DefaultResolutionProvider
 import org.ossreviewtoolkit.reporter.DefaultLicenseTextProvider
+import org.ossreviewtoolkit.reporter.HowToFixTextProvider
```
[the remark this answers is not in the document]
Comment (fviernau) [reply cli/src/main/kotlin/commands/ReporterCommand.kt:53]: Done.

File: cli/src/main/kotlin/commands/ReporterCommand.kt [hunk 8]
```diff
@@ -160,6 +161,14 @@ class ReporterCommand : CliktCommand(
         .file(mustExist = true, canBeFile = true, canBeDir = false, mustBeWritable = false, mustBeReadable = true)
         .convert { it.absoluteFile.normalize() }
 
+    private val howToFixTextProviderScript by option(
+        "--how-to-fix-text-provider-script",
+        help = "The path to a Kotlin script which returns an instance of a 'HowToFixTextProvider'. That provider can " +
+                "inject how-to-fix texts in Markdown formats for ORT issues."
```
[the remark this answers is not in the document]
Comment (fviernau) [reply cli/src/main/kotlin/commands/ReporterCommand.kt:167]: Done.

File: cli/src/funTest/kotlin/ExamplesTest.kt (hunk 2 above)
Comment (fviernau) [reply cli/src/funTest/kotlin/ExamplesTest.kt:126]: Done.

File: examples/how-to-fix-text-provider.kts (hunk 1 above)
Comment (fviernau) [reply examples/how-to-fix-text-provider.kts:8]: Done.

**Review by mnonnenmacher — DISMISSED** [review]

File: reporter/src/main/kotlin/utils/ReportTableModelMapper.kt [hunk 9]
```diff
@@ -85,7 +87,8 @@ class ReportTableModelMapper(
                 }
             },
             isResolved = resolutions.isNotEmpty(),
-            severity = severity
+            severity = severity,
+            howToFix = howToFixTextProvider.getHowToFixText(this) ?: ""
```
Comment (sschuberth) [inline reporter/src/main/kotlin/utils/ReportTableModelMapper.kt:91]: Nit: Could be `.orEmpty()`.

File: reporter/src/funTest/assets/static-html-reporter-test-expected-output.html [hunk 10]
```diff
@@ -536,6 +536,26 @@ <h3>Packages</h3>
                   <p>Unknown time [ERROR]: FileCounter - DownloadException: No source artifact URL provided for 'Gradle:org.ossreviewtoolkit.gradle.example:lib:1.0.0'.<br>Caused by: DownloadException: No VCS URL provided for 'Gradle:org.ossreviewtoolkit.gradle.example:lib:1.0.0'. Please make sure the release POM file includes the SCM connection, see: https://docs.gradle.org/current/userguide/publishing_maven.html#example_customizing_the_pom_file</p>
                   <p></p>
                 </li>
+                <details>
+                  <summary>How to fix</summary>
+                  <ul>
+
```
Comment (sschuberth) [inline reporter/src/funTest/assets/static-html-reporter-test-expected-output.html:542]: Can we drop the superfluous blank lines?

File: reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt [hunk 11]
```diff
@@ -383,6 +389,12 @@ class StaticHtmlReporter : Reporter {
                                 issueDescription(issue)
                                 p { +issue.resolutionDescription }
                             }
+                            if (!issue.isResolved && issue.howToFix.isNotBlank()) {
```
Comment (sschuberth) [inline reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt:392]: We usually use blank lines to separate code blocks.

File: examples/how-to-fix-text-provider.kts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+                | To fix this issue please proceed with the following steps:
+                |
+                |1. Manually verify that the file does not contain any license information.
+                |2. File an error resolution in the resolutions file of your configuration directory with a comment
+                |   stating that the file does not contain any license information.
+                |   ```
+                |
+            """.trimIndent()
+        }
+        return null
```
Comment (sschuberth) [inline examples/how-to-fix-text-provider.kts:17]: We usually use blank lines to separate from code blocks.

**Review by sschuberth — DISMISSED** [review]

File: reporter/src/funTest/assets/static-html-reporter-test-expected-output.html (hunk 10 above)
Comment (fviernau) [reply reporter/src/funTest/assets/static-html-reporter-test-expected-output.html:542]: @sschuberth I had to undo the removal of the superfluous blank lines.
I think it's ok to keep them in the context of this PR as the output is identical with the how-to-fix sections for `rule violations` in that same file.

Comment (sschuberth) [reply reporter/src/funTest/assets/static-html-reporter-test-expected-output.html:542]: > @sschuberth I had to undo the removal of the superfluous blank lines.
> I think it's ok to keep them in the context of this PR as the output is identical with the how-to-fix sections for `rule violations` in that same file.

Ok, but I'll file a follow-up issue to remove those lines from both places.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**